### PR TITLE
CSP-610-PopulateProfileDataInSessionModel

### PR DESCRIPTION
**Pull request by ravihejmadi**
This PR will not go out into Testing. It will be reviewed and pushed into main.
New ProfileService, GetProfilesResult, Profile.cs DomainModel 
Updated IOuterApiClient.cs, ServiceRegistrationsStartup.cs, BeforeYouStartController & Tests

### CODE REVIEW

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+[TestFixture]
+public class ProfileServiceTests
+{
+
+    [MoqAutoData]
+    public async Task Service_ProfileData_ReturnsProfiles(
+        [Frozen] Mock<IOuterApiClient> _outerApiClient)
+    {
+        const string userType = "apprentice";
+        var service = new ProfileService(_outerApiClient.Object);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs:19]: Rename service to sut so it is easy to identify the test target. 

Also there should be spaces between Arrange / Act / Assert for readability

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        var profiles = await service.GetProfilesByUserType(userType);
+
+        var result = await _outerApiClient.Object.GetProfilesByUserType(userType);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs:22]: This works but it is not the correct way of doing things. You should do a setup on the client mock to return a pre-defined result. Then assert that the response received from the service calls is equivalent to the pre-defined set.

File: src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs [hunk 3]
```diff
@@ -23,4 +23,5 @@ public class ProfileModel
 {
     public int Id { get; set; }
     public string Value { get; set; } = null!;
+    public string Category { get; set; } = null!;
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs:26]: You need to save description and order as well here.

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs [hunk 4]
```diff
@@ -32,9 +34,13 @@ public IActionResult Get()
     }
 
     [HttpPost]
-    public IActionResult Post()
+    public async Task<IActionResult> PostAsync()
     {
+        var profiles = await _profileService.GetProfilesByUserType("apprentice");
         _sessionService.Set(new OnboardingSessionModel());
+        var sessionModel = _sessionService.Get<OnboardingSessionModel>();
+        sessionModel.ProfileData = profiles.ConvertAll(p => new ProfileModel());
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs:42]: The conversion of data is not happening here, you will just get a list of empty objects. 
You need an implicit operator on ProfileModel to convert from Profile and then you can convert the list using select `profiles.Select(p => (ProfileModel)p).ToList()`

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs [hunk 5]
```diff
@@ -13,13 +13,51 @@ namespace SFA.DAS.ApprenticeAan.Web.UnitTests.Controllers.Onboarding.BeforeYouSt
...
+        _profileService.Verify(p => p.GetProfilesByUserType(userType));
+
+        sessionServiceMock.Setup(s => s.Get<OnboardingSessionModel>()).Returns(sessionModel);
+        sessionServiceMock.Object.Set(sessionModel);
+
+        await sut.PostAsync();
+
+        sessionServiceMock.Verify(s => s.Set(sessionModel));
+
+        sessionModel.ProfileData.Should().BeEquivalentTo(profiles.ConvertAll(p => new ProfileModel()));
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs:46]: There should not be a need to do the conversion here at all. I have added a comment on the actual code that will help remove any conversion in test.

**Review by DevenKShah — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs (hunk 1 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs]: Renamed service & Added new lines in-between

File: src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs (hunk 3 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs:29]: Added Description and Ordering

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs (hunk 4 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs]: Modified this line and added an implicit conversion in the model [here](https://github.com/SkillsFundingAgency/das-apprentice-aan-web/blob/d07e4577a0a4b0f597816714a3e8a9da60ec7eab/src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs#L32)

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs (hunk 5 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs]: Amended this line

File: src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs (hunk 2 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Application.UnitTests/Services/ProfileServiceTests.cs]: This Unit Test has now been reworked. Thanks for you help with this.

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs [hunk 6]
```diff
@@ -13,13 +13,51 @@ namespace SFA.DAS.ApprenticeAan.Web.UnitTests.Controllers.Onboarding.BeforeYouSt
...
+         [Greedy] BeforeYouStartController sut)
+    {
+        OnboardingSessionModel sessionModel = new();
+
+        sessionServiceMock.Setup(s => s.Get<OnboardingSessionModel>()).Returns(sessionModel);
+
+        var result = await sut.PostAsync();
+
+        result.As<RedirectToRouteResult>().Should().NotBeNull();
+        result.As<RedirectToRouteResult>().RouteName.Should().Be("TermsAndConditions");
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs:61]: User the RouteNames constant here

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs [hunk 7]
```diff
@@ -13,13 +13,51 @@ namespace SFA.DAS.ApprenticeAan.Web.UnitTests.Controllers.Onboarding.BeforeYouSt
...
+
+        sessionServiceMock.Setup(s => s.Get<OnboardingSessionModel>()).Returns(sessionModel);
+        sessionServiceMock.Object.Set(sessionModel);
+
+        await sut.PostAsync();
+
+        sessionServiceMock.Verify(s => s.Set(sessionModel));
+
+        sessionModel.ProfileData.Should().BeEquivalentTo(profiles.Select(p => (ProfileModel)p).ToList());
+    }
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs:47]:
```suggestion (fix any compile time issues)
    [MoqAutoData]
    public async Task Post_GetProfiles_UpdatesSessionModel(
        [Frozen] Mock<ISessionService> sessionServiceMock,
        [Frozen] Mock<IProfileService> profileServiceMock,
        [Greedy] BeforeYouStartController sut),
        List<Profile> profiles
    {
        const string userType = "apprentice";

        OnboardingSessionModel sessionModel = new();
        profileServiceMock.Setup(s => s..GetProfilesByUserType(userType)).ReturnsAsync(profiles);
        sessionServiceMock.Setup(s => s.Get<OnboardingSessionModel>()).Returns(sessionModel);
               
        await sut.PostAsync();

        sessionServiceMock.Verify(s => s.Set(sessionModel));

        sessionModel.ProfileData.Should().BeEquivalentTo(profiles));
    }
```

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs [hunk 8]
```diff
@@ -32,9 +34,13 @@ public IActionResult Get()
     }
 
     [HttpPost]
-    public IActionResult Post()
+    public async Task<IActionResult> PostAsync()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs:37]: Rename to `Post()`

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs [hunk 9]
```diff
@@ -32,9 +34,13 @@ public IActionResult Get()
...
     {
+        var profiles = await _profileService.GetProfilesByUserType("apprentice");
         _sessionService.Set(new OnboardingSessionModel());
+        var sessionModel = _sessionService.Get<OnboardingSessionModel>();
+        sessionModel.ProfileData = profiles.Select(p => (ProfileModel)p).ToList();
+        _sessionService.Set(sessionModel);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs:43]:
```suggestion
        var sessionModel = new OnboardingSessionModel();
        sessionModel.ProfileData = profiles.Select(p => (ProfileModel)p).ToList();
        _sessionService.Set(sessionModel);
```

File: src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs [hunk 10]
```diff
@@ -23,4 +25,15 @@ public class ProfileModel
 {
     public int Id { get; set; }
     public string Value { get; set; } = null!;
-}
+    public string Description { get; set; } = null!;
+    public string Category { get; set; } = null!;
+    public int Ordering { get; set; }
+
+    public static implicit operator ProfileModel(Profile profile) => new()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs:32]: Move ProfileModel into its own file and add coverage please.

**Review by DevenKShah — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs (hunk 6 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs]: Substituted with RouteNames constant

File: src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs (hunk 7 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web.UnitTests/Controllers/Onboarding/BeforeYouStartControllerTests/BeforeYouStartControllerPostTests.cs]: Applied profileServiceMock.Setup & removed LINQ on last line. Compiler null error message has gone. Thanks :-)

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs (hunk 8 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs]: Renamed to Post

File: src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs (hunk 9 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Controllers/Onboarding/BeforeYouStartController.cs:43]: @DevenKShah, I did this but then had to roll this back as [this](https://github.com/SkillsFundingAgency/das-apprentice-aan-web/pull/11#discussion_r1137440549) change would not work. Happy to discuss

File: src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs (hunk 10 above)
Comment (ravihejmadi) [reply src/SFA.DAS.ApprenticeAan.Web/Models/OnboardingSessionModel.cs]: Moved to its own class and added coverage

**Pull request merged**